core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/QueryJoinOptimizer.java

```diff
@@ -439,6 +443,23 @@ protected double getTupleExprCost(TupleExpr tupleExpr, Map<TupleExpr, Double> ca
 				}
 			}
 
+			// BindingSetAssignment has a typical constant cost. This cost is not based on statistics so is much more
+			// reliable. If the BindingSetAssignment binds to any of the other variables in the other tuple expressions
+			// to choose from, then the cost of the BindingSetAssignment should be set to 0 since it will always limit
+			// the upper bound of any other costs. This way the BindingSetAssignment will be chosen as the left
+			// argument.
+			if (tupleExpr instanceof BindingSetAssignment) {
```

Comment:
I get the idea but this should be solved in the EvaluationStatistics implementation. In fact, its default implement currently just assigns BindingSetAssignment a fixed cost of 1. Are you seeing any real cases where that leads to a BindingSetAssignment _not_ being optimally planned?

Reply:
Ah, hang on, I see you explained this on the issue ticket, l'll read